### PR TITLE
Zero alloc for empty IEnumerator<T>

### PR DESCRIPTION
**Pull request by benaadams**
Contributes to: "Reduce IEnumerator allocation via interface" https://github.com/dotnet/coreclr/issues/4446

### CODE REVIEW

File: src/mscorlib/shared/System/Collections/Generic/EmptyEnumerator.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+        public void Dispose() { }
+
+        DictionaryEntry IDictionaryEnumerator.Entry { get; } = default;
+
+        object IDictionaryEnumerator.Key { get; } = null;
+
+        object IDictionaryEnumerator.Value { get; } = null;
+
+        object IEnumerator.Current { get; } = null;
```
Comment (svick) [inline src/mscorlib/shared/System/Collections/Generic/EmptyEnumerator.cs]: Why are all these properties autoproperties that load their value from a field, when they could just directly return it? It's going to result in shorter source code and it should also be more efficient (though probably not significantly, unless the CLR gets better at devirtualization).

E.g. change the last one to:

```c#
object IEnumerator.Current => null;
```

**Comment by jkotas** [comment]
This is replacing (rare?) allocation of empty enumerator with unconditional allocation of a type. Do you have any data that this a good tradeoff?

File: src/mscorlib/shared/System/Collections/Generic/EmptyEnumerator.cs (hunk 1 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/EmptyEnumerator.cs]: Fixed

**Comment by benaadams** [comment]
Perhaps a bit big :-/
```
Total bytes of diff: 10834 (0.30% of base)
    diff is a regression.

Total byte diff includes 8405 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had   18 unique methods,     8405 unique bytes

Top file regressions by size (bytes):
       10834 : System.Private.CoreLib.dasm (0.30% of base)

1 total files with size differences (0 improved, 1 regressed), 0 unchanged.

Top method regessions by size (bytes):
        3405 : System.Private.CoreLib.dasm - EmptyEnumerator`1:InitializeShared():ref (0/54 methods)
        2277 : System.Private.CoreLib.dasm - EmptyEnumerator`1:get_Shared():ref (0/54 methods)
        1494 : System.Private.CoreLib.dasm - Dictionary`2:System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<TKey,TValue>>.GetEnumerator():ref:this (29 methods)
         827 : System.Private.CoreLib.dasm - List`1:System.Collections.Generic.IEnumerable<T>.GetEnumerator():ref:this (27 methods)
         757 : System.Private.CoreLib.dasm - EmptyEnumerator`1:get_Current():struct:this (0/39 methods)

26 total methods with size differences (0 improved, 26 regressed), 17255 unchanged.
```

**Comment by benaadams** [comment]
Probably not worth it

**Comment by jkotas** [comment]
The polymorphic iterator would introduce regressions in cases that benefit of the monomorphic iterator today. For example:

```csharp
static int Main()
{
    IDictionary<int,int> d = new Dictionary<int,int>();
    d.Add(1,1);

    int sum = 0;
    int start = Environment.TickCount;
    for (int i = 0; i < 100000000; i++)
    {
        foreach (var e in d)
            sum += (e.Key + e.Value);
    }   
    int end = Environment.TickCount;
    Console.WriteLine((end-start).ToString());
    return sum;
}
```

2891ms before the change, 3703ms with this change.

**Comment by benaadams** [comment]
> 2891ms before the change, 3703ms with this change.

Definitely not worth it

**Pull request closed without merging**